R/doubletFinder.R

```diff
@@ -87,13 +97,151 @@ doubletFinder <- function(seu, expected.doublets = 0, proportion.artificial = 0.
+  doublet_mat <- matrix(0, n_doublets, length(genes))
+  if (n_doublets <= 50) {
+    doublet_mat <-
+      (obj[["matrix"]][real_ind1,] + obj[["matrix"]][real_ind2,]) / 2
+  } else {
+    n_chunks <- floor(n_doublets / 50) # the last chunk is added separately
+    pb <- txtProgressBar(style = 3)
+    for (i in 1:n_chunks) {
+      ind1 <- 1 + (i - 1) * 50
+      ind2 <- i * 50
+      doublet_mat[ind1:ind2,] <-
+        (obj[["matrix"]][real_ind1[ind1:ind2],] + obj[["matrix"]][real_ind2[ind1:ind2],]) / 2
+      setTxtProgressBar(pb, i / n_chunks)
+    }
+    close(pb)
+    remainder <- n_chunks %% 50
+    if (remainder > 0) {
+      doublet_mat[(n_chunks * 50 + 1):n_doublets,] <-
+        (obj[["matrix"]][real_ind1[(ind2 + 1):n_doublets],] + obj[["matrix"]][real_ind2[(ind2 + 1):n_doublets],]) / 2
+    }
+  }
+  print("Writing artificial doublets to loom file")
+  doublets <- create_doublet_loom(n_doublets, doublet_mat, obj, overwrite = overwrite)
+  # Free up memory
+  rm(doublet_mat)
+  gc(verbose = FALSE)
+  # Combine real cells and doublets
+  print("Combining real cells and doublets")
+  loom_wdoublets <- loomR::combine(list(obj, doublets), "loom_wdoublets.loom",
+                                    overwrite = overwrite)
+  gc(verbose = FALSE)
+  ## Step 2: Pre-process real-artificial merged data using Seurat
+  print("Normalizing Seurat object...")
+  Seurat::NormalizeData(loom_wdoublets, overwrite = TRUE,
+                scale.factor = as.integer(hdf5r::h5attributes(obj[["layers/norm_data"]])$calc_params[1]),
+                chunk.size = chunk.size, chunk.dims = chunk.dims)
+  print("Finding variable genes...")
+  fvg_params <- as.numeric(hdf5r::h5attributes(obj[["row_attrs/var_genes"]])$calc_params[1:4])
+  Seurat::FindVariableGenes(loom_wdoublets, overwrite = TRUE,
+                            x.low.cutoff = fvg_params[1],
+                            x.high.cutooff = fvg_params[2],
+                            y.cutoff = fvg_params[3],
+                            y.high.cutoff = fvg_params[4],
+                            chunk.size = chunk.size)
+  print("Scaling data...")
+  Seurat::ScaleData(loom_wdoublets, display.progress = TRUE, overwrite = TRUE,
+                    chunk.size = chunk.size, chunk.dims = chunk.dims)
+  print("Running PCA...")
+  # Get the names of the highly variable genes
+  var_gene_inds <- loom_wdoublets[["row_attrs/var_genes"]][]
+  var_genes <- genes[var_gene_inds]
+  Seurat::RunPCA(loom_wdoublets, pcs.compute = pcs.compute, overwrite = TRUE,
+                 pc.genes = var_genes,
+                 do.print = FALSE,
+                 chunk.size = chunk.size, ...)
+
+  ## Step 4: Create distance matrix from PCA results
+  print("Calculating PC distance matrix...")
+  # This will be loaded into memory; since the number of PCs computed is small,
+  # this shouldn't be too bad
+  pca.coord <- t(loom_wdoublets$col.attrs$pca_cell_embeddings[,])
+  dist.mat <- as.matrix(dist(pca.coord))
```

Comment:
We will be releasing a doubletFinder update in the next couple of weeks that uses rdist instead of dist, which is much faster

Reply:
OK